### DISCUSSION TITLE
Statistics generation and analysis (Hibernate, Infinispan)

### DISCUSSION THREAD

**Discussion by dm-vestmark** [infinispan/infinispan#11650] [Q&A]
We would like some guidance on two issues pertaining to hibernate statistics in a clustered environment.
- How to generate statistics for a node in the cluster and not cluster-wide
- How to use these statistics to understand trends in cache usage

We are looking into an apparent sporadic cache performance degradation over a period of several days in a production environment. We are able to generate statistics at the cluster level using the following additions to our config files:

standalone-ha.xml :
`<cache-container` name="VmManagedBatchedMda" marshaller="JBOSS" modules="org.infinispan.hibernate-cache" `statistics-enabled="true">`
`<invalidation-cache` name="entity" remote-timeout="120000" `statistics-enabled="true">`
`<heap-memory` `size="100"/>` <expiration max-idle="300000" `lifespan="300000"/>`
`</invalidation-cache>`

and setting hibernate.cache.infinispan.statistics to true.
"[hibernate.cache.infinispan.statistics]": true

How to generate statistics for a node in the cluster and not cluster-wide
These additions have produced only statistics at the cluster level and not the local node level. The cluster-wide statistics do not allow us to analyze why a specific node(s) experience performance degradation based on cache usage. Thus, we need to generate and analyze statistics at the node level.
When we do this in a non-clustered environment (with standalone.xml and a ), we get a properly working Statistics mbean for each Hibernate entity. We get a properly working ClusterCacheStats bean in a cluster, but the values are the same on each cluster node. The Statistics mbean is still there, but we get an org.infinispan.commons.CacheException: java.lang.NullPointerException trying to access every single JMX attribute from the Statistics bean. Are the local Statistics supposed to work for an invalidation cache and report the local statistics (hits, misses, evictions) for each cluster node? And how does the ClusterCacheStats aggregate the numbers – is it adding up the hits, misses, evictions, and invalidations across all cluster nodes and reporting the sum?
How to use these statistics to understand trends in cache usage
The statistics also need to be captured for usage analysis over time. Is there some easy way to log the statistics over time for multiple entities? Or is the only way to configure our JMX tools to capture and track multiple attributes for multiple mbeans corresponding to the various entities we are interested in?
The Environment: Wildfly version: 26.0.1.Final
Inifinispan version: 12.1.7.Final
Hibernate version: 5.4.18.Final

**Comment by dm-vestmark** [comment]
Hi Community, any assistance would be much appreciated

**Reply by vmgbrown** [reply]
FWIW, this is a more detailed version of a similar question I posted several months ago:
https://github.com/infinispan/infinispan/discussions/10933

**Comment by tristantarrant** [comment]
@pferraro can you help investigating this ?

**Reply by dm-vestmark** [reply]
Hi Paul, any assistance would be very much appreciated

**Reply by pferraro** [reply]
Do you have a stacktrace containing the NPE when viewing the "Statistics" mbean?

**Reply by dm-vestmark** [reply]
We will review the log and provide if we have one.

**Reply by vmgbrown** [reply]
[jmx-log.txt](https://github.com/infinispan/infinispan/files/13879870/jmx-log.txt)

**Reply by dm-vestmark** [reply]
Log provided in the jmx-log.txt

**Reply by tristantarrant** [reply]
Are you deploying the hawt.io war file ? If so, what version is it ?

**Reply by tristantarrant** [reply]
Add the following line:
```
<module name="org.eclipse.microprofile.metrics.api"/>
```
to `modules/system/layers/base/org/infinispan/main/module.xml`

and use Hawt.io 3.0

**Reply by tristantarrant** [reply]
@dm-vestmark did you try this?

**Comment by tristantarrant** [comment]
Would it be possible for you to try with a more recent WildFly ?

**Reply by dm-vestmark** [reply]
What version would you suggest?

**Reply by tristantarrant** [reply]
At least 28

**Reply by tristantarrant** [reply]
Unfortunately WildFly >= 27 are JakartaEE only, so that may not be a valid upgrade for you guys.

**Reply by dm-vestmark** [reply]
Let's go back a bit and simplify the question.  
Our Environment: 

Wildfly version: 26.0.1.Final
Inifinispan version: 12.1.7.Final
Hibernate version: 5.4.18.Final

We want node-level statistics in addition to cluster wide.  
It must have worked at some point.

Please assist.

**Reply by tristantarrant** [reply]
The problem is because statistics in Infinispan 12 rely on Microprofile Metrics but WildFly is not exposing those classes to the `org.infinispan:main` module thus causing a `ClassNotFoundException`

**Reply by vmgbrown** [reply]
@tristantarrant Is that a limitation of WildFly 26? Are those classes exposed by WildFly 27 or later?

**Reply by tristantarrant** [reply]
It's a bug with WildFly 26. You can patch it as I described in a comment above. The bug is also present in all subsequent releases of WildFly. However, I doubt you could upgrade anyway, since WildFly 27+ require JakartaEE 9 deployments and don't work with JavaEE deployments.

**Reply by vmgbrown** [reply]
I see. So the suggestion to modify module.xml is the patch. Is Hawtio 3.0 also required? We have not been able to get it to work with WildFly 26.

**Reply by tristantarrant** [reply]
I simply deployed the hawtio 3.0 war to WF 26 with no issues. If older hawt.io works for you, that's fine.
